### PR TITLE
CMake jit_preprocess_files function only runs when needed

### PR DESCRIPTION
**Pull request by robertmaynard**
This correct a couple of issues with the `jit_preprocess_files` function.

1. It now always makes sure that the output directory that the jit headers are going into will exist before executing. So if a user deletes `<build_dir>/include` everything will work as expected
2. The output of the jitify tool are files that end in `.jit.hpp` not `.jit`. This is required to allow generators to build the minimal incremental compilation graph
3. Mark the input source files as a dependency so when that file is modified, the build-system knows to recompile the jit file.

### CODE REVIEW

**Review by cwharris — APPROVED** [review]

**Comment by cwharris** [comment]
Since `ARG_OUTPUT` is used as both the explicit `OUTPUT` and to build a list of dependencies, this change doesn't effect the dependency tree, because custom commands are run _every_ build, yes?

**Comment by ttnghia** [comment]
I applied this PR but that still does not help---I still get the same error:
```
cpp/src/binaryop/binaryop.cpp:23:10: fatal error: jit_preprocessed_files/binaryop/jit/kernel.cu.jit.hpp: No such file or directory
   23 | #include <jit_preprocessed_files/binaryop/jit/kernel.cu.jit.hpp>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

**Comment by robertmaynard** [comment]
> I applied this PR but that still does not help---I still get the same error:
> 
> ```
> cpp/src/binaryop/binaryop.cpp:23:10: fatal error: jit_preprocessed_files/binaryop/jit/kernel.cu.jit.hpp: No such file or directory
>    23 | #include <jit_preprocessed_files/binaryop/jit/kernel.cu.jit.hpp>
>       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated.
> ```

If you run `make jitify_preprocess_run` ( or `ninja jitify_preprocess_run` ) do you see that the file `<build_dir>/include/ jit_preprocessed_files/binaryop/jit/kernel.cu.jit.hpp` exists?

**Comment by ttnghia** [comment]
I just check that. The folder `include/ jit_preprocessed_files` is empty---there is not any subfolder.

**Comment by ttnghia** [comment]
FYI, I tested this PR by deleting everything and checked out the latest branch-0.20 and did a fresh build.

**Comment by robertmaynard** [comment]
> Since `ARG_OUTPUT` is used as both the explicit `OUTPUT` and to build a list of dependencies, this change doesn't effect the dependency tree, because custom commands are run _every_ build, yes?

Hmmm. I think you are correct as is they will be recompiled each time. But we can fix that to only re-compile when the input file is modified.

**Comment by robertmaynard** [comment]
> I just check that. The folder `include/jit_preprocessed_files` is empty---there is not any subfolder. Nothing there.

I can reproduce this behavior on Ubuntu 20.04 with the Makefile Generator, while the ninja generator correctly generates the output file.

**Comment by ttnghia** [comment]
> > I can reproduce this behavior on Ubuntu 20.04 with the Makefile Generator but ninja generator correctly generated.

Sorry that I mistakenly modified your post---I intentionally wanted to reply-quote instead.

My system is exactly like that: Ubuntu 20.04 with the Makefile Generator.

**Comment by ttnghia** [comment]
Your new commits fix the issue. Yeah.

**Review by ttnghia — APPROVED** [review]

**Comment by robertmaynard** [comment]
@cwharris 

This now fixes the issue where builds with the `Unix Makefiles` generator not creating the output directory. It also fixes a couple of issues allowing for better incremental builds. Now CMake will only re-run the JIT step when the input files change.  Previously it always was running the JIT step as the output files listed in `OUTPUT`  didn't exist ( either bad names or due to the above generator problem ).

**Comment by cwharris** [comment]
Thanks @robertmaynard !

I didn't see this locally because ninja was generating the files whereas cmake does not.

**Review by jlowe — APPROVED** [review]

File: cpp/cmake/Modules/JitifyPreprocessKernels.cmake [hunk 1]
```diff
@@ -16,8 +16,6 @@
 
 cmake_minimum_required(VERSION 3.18)
```
Comment (trxcllnt) [inline cpp/cmake/Modules/JitifyPreprocessKernels.cmake:17]: Need to remove this.

File: cpp/cmake/Modules/JitifyPreprocessKernels.cmake [hunk 2]
```diff
@@ -16,8 +16,6 @@
...
 
-file(MAKE_DIRECTORY "${CUDF_GENERATED_INCLUDE_DIR}/include/jit_preprocessed_files")
-
 # Create `jitify_preprocess` executable
 project(jitify_preprocess VERSION 2.0 LANGUAGES CXX CUDA)
```
Comment (trxcllnt) [inline cpp/cmake/Modules/JitifyPreprocessKernels.cmake:20]: Need to remove this line too.

Comment (cwharris) [reply cpp/cmake/Modules/JitifyPreprocessKernels.cmake:20]: we're `include`ing this file in cudf's CMakeLists, so this actually ends up setting cudf's version, which we don't want. :disappointed:

**Review by kkraus14 — APPROVED** [review]

**Comment by kkraus14** [comment]
@gpucibot merge

**Comment by kkraus14** [comment]
rerun tests

**Comment by ttnghia** [comment]
Rerun tests.

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
FYI -- ops team is building a new `rapids-build-env` package that has updated conda packages. The previous package was built at a perfect timing in between the new `fsspec` and `gcsfs` versions being released.

**Comment by kkraus14** [comment]
rerun tests

**Comment by kkraus14** [comment]
@gpucibot merge

**Pull request merged**